index.js

```diff
@@ -580,6 +593,150 @@ module.exports = {
-    }
+    };
+
+    self.getPagesTree = async (req) => {
+      self.maps = {};
+      const excludedTypes = [
+        'workflow-document',
+        ...self.options.excludeTypes || []
+      ]
+
+      const pages = await getPages();
+      const pagesWithPieces = await getPieces(pages);
+      const pagesTree = buildPagesTree(pagesWithPieces);
+
+      return pagesTree;
+
+      async function getPages () {
+        try {
+          const pages = await self.apos.pages
+            .find(req, {}, {
+              _id: 1,
+              title: 1,
+              slug: 1,
+              path: 1,
+              level: 1,
+              rank: 1,
+              _url: 1
+            })
+            .areas(false)
+            .joins(false)
+            .sort({
+              level: -1,
+              rank: -1
+            })
+            .toArray();
+
+          return pages
+            .filter((page) => !excludedTypes.includes(page.type))
+            .map((page) => ({
+              ...page,
+              _children: []
+            }));
+        } catch (err) {
+          self.apos.utils.error(err);
+        }
+      }
+
+      async function getPieces (pages) {
+        const piecesModules = self.getPiecesModules();
+
+        for (const mod of piecesModules) {
+          if (excludedTypes.includes(mod.name)) {
+            continue;
+          }
+
+          const modulePieces = [];
+          await fetchPieces(req, {
+            mod,
+            skip: 0,
+            modulePieces
+          });
+
+          modulePieces.forEach((piece) => {
+            if (!piece._url || pages.some((page) => page._url === piece._url)) {
+              return;
+            }
+
+            if (!piece._parentUrl) {
+              pages.push(piece);
+            } else {
+              pages.forEach((page) => {
+                if (page._url === piece._parentUrl) {
+                  page._children.push(piece);
+                }
+              });
+            }
+          });
+        }
+
+        return pages;
+
+        async function fetchPieces (req, {
+          mod, skip, modulePieces
+        }) {
+          try {
+            const fetchedPieces = await self.findPieces(req, mod, { _id: 1, title: 1, _url: 1 })
+              .skip(skip)
+              .limit(self.piecesPerBatch)
+              .toArray();
+
+
+            if (!Array.isArray(fetchedPieces)) {
+              return;
+            }
+
+            fetchedPieces.forEach(piece => {
+              if (piece._url && !excludedTypes.includes(piece.type)) {
+                modulePieces.push(piece);
+              }
+            });
+
+            if (fetchedPieces.length) {
+              await fetchPieces(req, {
+                mod,
+                skip: skip + fetchedPieces.length,
+                modulePieces
+              });
+            }
+
+          } catch (err) {
+            self.apos.utils.error(err);
+          }
+        }
+      }
+
+      function buildPagesTree (pages) {
```

Comment:
Could have made this function cleaner by reducing but already did this function for UC, so I got it. Didn't want to spend too much time on this.